Data/Vector/Generic.hs

```diff
@@ -241,8 +241,13 @@ infixl 9 !?
 -- | O(1) Safe indexing
 (!?) :: Vector v a => v a -> Int -> Maybe a
 {-# INLINE_FUSED (!?) #-}
-v !? i | i < 0 || i >= length v = Nothing
-       | otherwise              = Just $ unsafeIndex v i
+-- Lengths are never negative, so we can check  0 <= i < length v
+-- using one unsigned comparison.
```

Comment:
I don't understand this comment and logic. Is the idea that 

```
fromIntegral (minBound :: Int) :: Word
9223372036854775808
```

is always bigger than any realistic length `Vector` could be?

I think that should explained more clearly, this is "too clever" otherwise.

Reply:
@phadej, right; a vector of length over `maxBound :: Int` will break assumptions all over the place. I can add a comment to that effect.

Reply:
Existing definition has a benefit of short-circuiting in compile time when `i` is a statically known negative number. But I do not think that this is an important scenario.

Reply:
It's probably worth adding note that `fromIntegral (i :: Int) :: Word > fromIntegral (maxBound :: Int) :: Word` for all negative Ints. Such optimization is not entirely obvious